In [1]:
# initializing otter-grader
import otter
grader = otter.Notebook()

# Homework 2: Sampling Error and Bias
## Due Date:  April 24, 11:59PM

## Collaboration Policy

Data science is a collaborative activity. While you may talk with others about
the homework, we ask that you **write your solutions individually**. If you do
discuss the assignments with others please **include their names** below.

**Collaborators**: None

## Grading

Grading is broken down into autograded answers and free response. 

For autograded answers, the results of your code are compared to provided and/or hidden tests. **For autograded probability questions, the provided tests will only check that your answer is within a reasonable range.**

For free response, readers will evaluate how well you answered the question and/or fulfilled the requirements of the question.

For plots, make sure to be as descriptive as possible: include titles, axes labels, and units wherever applicable.

**Please make sure to save regularly; do not rely on autosave to save your work for you!**

## Introduction

The outcome of the US presidential election in 2016 took many people and many pollsters by surprise. In this assignment we will carry out a simulation study / post mortem in an attempt to understand what happened.

In [2]:
import numpy as np
import pandas as pd

from IPython.display import display, Latex, Markdown

## Plotting functions
import altair as alt
alt.data_transformers.disable_max_rows()

DataTransformerRegistry.enable('default')

### Background

In Pennsylvania, 6,165,478 people voted in the election.
Trump received 48.18% of the vote and Clinton recieved 47.46%.
This doesn't add up to 100% because other candidates received votes.
All together these other candidates received 100% - 48.18% - 47.46% = 4.36% of the vote.

Suppose we could select one person at random from the 6+ million voters in PA (note: PA is the common abbreviation for Pennsylvania). We are interested in the chance that we'd choose a Trump, Clinton, or Other voter.

Below is a probability table for the choice:


|   Voted for   |  Trump|    Clinton|    Other|
|-----------|-----------|-----------|---------|
| Probability      |   0.4818   | 0.4746  |   0.0436 |
| Number of people | 2,970,733  | 2,926,441 | 268,304 |

### Question 1

Suppose we take a simple random sample of $n = 1500$ voters from the 6+ million voters in PA.

What is the expected number of Trump voters?
What is the expected number of Clinton voters?

To answer these questions, let $T_1$ be 1 if the first voter chosen for the sample voted for Trump and 0 if they voted for Clinton or another candidate.

Let $T_2$ be 1 if the second voter chosen for the sample voted for Trump and 0 if they voted for Clinton or another candidate, and so on.

Let's start with some very basic questions. Find: 
$$P(T_{1000} = 1)$$

$$P(T_{1000} = 0)$$

$$\mathbb{E}(T_{17})$$






#### Part 1 ####
$P(T_{1000} = 1)$:

<!--
BEGIN QUESTION
name: q1a
manual: true
points: 3
-->
<!-- EXPORT TO PDF -->

In [3]:
q1a_answer = 0.4818
q1a_answer

0.4818

#### Part 2 ####
$P(T_{1000} = 0)$:

<!--
BEGIN QUESTION
name: q1b
manual: true
points: 3
-->
<!-- EXPORT TO PDF -->

In [4]:
q1b_answer = 0.4746 + 0.0436
q1b_answer

0.5182

#### Part 3 ####
$E(T_{17})$:

<!--
BEGIN QUESTION
name: q1c
manual: true
points: 3
gradescope: show
-->
<!-- EXPORT TO PDF -->

In [5]:
q1c_answer = 1 * 0.4818 + 0 * 0.5182
q1c_answer

0.4818

### Question 2

For each voter chosen for the sample, let's keep track of whether they voted for Trump, Clinton, or another candidate.

We can do this with the triple $(T_i, C_i, O_i)$, $i = 1, 2, \ldots, 1500$, where 

$T_i = 1$ if the $i$th voter sampled voted for Trump and 0 otherwise (this is the same random variable as in Question 1), <br>
$C_i = 1$ if the $i$th voter sampled voted for Clinton and 0 otherwise, and <br>
$O_i = 1$ if the $i$th voter sampled voted for another candidate and 0 otherwise.

#### Part 1 ####

Find

$$ T_{17} + C_{17} + O_{17} = ?$$

<!--
BEGIN QUESTION
name: q2a
manual: true
points: 3
-->
<!-- EXPORT TO PDF -->

In [6]:
q2a_answer = 1
q2a_answer

1

#### Part 2 ####

Define $N_T = \sum_{i=1}^{1500} T_i$ and $N_C = \sum_{i=1}^{1500} C_i$, and
$N_O = \sum_{i=1}^{1500} O_i$

Notice that because they are sums of random variables, $N_T$ and $N_C$ and $N_O$ are random variables, too.

Find the expected value of $N_T$, i.e.,

$$\mathbb{E}(N_T)$$

In other words, find the expected number of Trump voters in our simple random sample of 1500 voters from Pennsylvania.

<!--
BEGIN QUESTION
name: q2b
manual: true
points: 3
-->
<!-- EXPORT TO PDF -->

In [7]:
q2b_answer = 1500 * 0.4818
q2b_answer

722.7

#### Part 3 ####
Find
$$N_T + N_C + N_O = ?$$

<!--
BEGIN QUESTION
name: q2c
manual: true
points: 3
-->
<!-- EXPORT TO PDF -->

In [8]:
q2c_answer = 1500
q2c_answer

1500

### Question 3

Given our population of Pennsylvania voters, every possible SRS has a certain well defined probability of occurring. For example, if we collected a SRS of only 2 voters without replacement (instead of 1500), the chance of each SRS is given in the probability distribution table below:

|   $N_T$   |  $N_C$    |    $N_O$  |    $p$    |
|-----------|-----------|-----------|-----------|
|    0      |   0       |      2    |   0.00189373515   |
|    0      |   1       |      1    |   0.04131080160    |
|    1      |   0       |      1    |   0.04193604504   |
|    0      |   2       |      0    |   0.22529213625   |
|    1      |   1       |      0    |   0.45740422268   |
|    2      |   0       |      0    |   0.23216338697   |



As an exercise in probability, we will have you compute similar probabilities for a simple random sample of 4 people without replacement.

#### Part 1 ####

Find the following probability. Give your answer as an exact number (i.e. as a product of fractions).

$$ P(N_T = 4, N_C = 0, N_O = 0)$$

Hint: It is a product of four fractions, each of which has a distinct numerator and denominator.

<!--
BEGIN QUESTION
name: q3a
manual: true
points: 2
-->
<!-- EXPORT TO PDF -->

$\frac{723}{1500}\frac{722}{1499}\frac{721}{1498}\frac{720}{1497}$

#### Part 2 ####

The answer from part 1 was a bit unwieldy.

We can simplify the problem by assuming that the draws are *with replacement*. That is, once a draw is picked as a sample, it can be picked again in the future. In this case, what is the following probability:

$$ P(N_T = 4, N_C = 0, N_O = 0)$$

Give your answer as an exact value.

<!--
BEGIN QUESTION
name: q3b
manual: true
points: 2
-->
<!-- EXPORT TO PDF -->

$\frac{4!}{4!0!0!}*(0.4818^{4}0.4746^{0}0.0436^{0})$

#### Part 3 ####

Under this same simplfying assumption (that we sample with replacement), find the following probabilities:

$$ P(N_T = 2, N_C = 2, N_O = 0)$$
$$ P(N_T = 2, N_C = 1, N_O = 1)$$

<!-- Hint: See the "Fun Problem Related to HW" from the [lecture]() if you're not sure how to proceed. -->

<!-- EXPORT TO PDF -->

$\frac{4!}{2!2!0!}*(0.4818^{2}0.4746^{2}0.0436^{2})$ 

$\frac{4!}{2!1!1!}*(0.4818^{2}0.4746^{1}0.0436^{1})$

### Question 4

Can you generalize the above probability calculation and
express the probability in terms of a probability mass function?

To set up the problem, let random variables $N_1, N_2, N_3$ be the number of Trump, Clinton, and Other voters selected, respectively.

Let $p_1, p_2, p_3$ be the chance of a Trump, Clinton, Other voter being chosen, respectively,
and let $n$ be the size of the sample drawn (with replacement).

In general, what is

$$ P(N_1 = k_1, N_2 = k_2, N_3 = k_3) = ?$$

Hint: The answer involves $n!, k_1!, k_2!, k_3!$ and $p_1, p_2, p_3$ raised to various powers. Also note this may be a particularly tough problem depending on your math background. Take your time, and please discuss with fellow students and instructors. You don't need to get this problem right to complete the questions later in this homework.

<!-- EXPORT TO PDF -->

$\frac{n!}{k_1!k_2!k_3!}*(p_1^{k_1}p_2^{k_2}p_3^{k_3})$

Fun fact: you have just derived the probabilty mass function of the multinomial distribution!

## Election Polling

Political polling is a type of public opinion polling that can at best represent a snapshot of public opinion at the particular moment in time. Voter opinion shifts from week to week, even day to day, as candidates battle it out on the campaign field.

Polls usually start with a "horse-race" question, where respondents are asked whom they would vote for in a head-to-head race if the election were tomorrow: Candidate A or Candidate B. The survey begins with this question so that the respondent is not influenced by any of the other questions asked in the survey. Some of these other questions are asked to help assess how likely is it that the respondent will vote. Other questions are asked about age, education, and sex in order to adjust the findings if one group appears overly represented in the sample.

Pollsters typically use [random digit dialing](https://en.wikipedia.org/wiki/Random_digit_dialing) to contact people.

### Question 5

#### Part 1 ####

If we're trying to predict the results of the Clinton vs. Trump presidential race, what is the population of interest? 

<!-- EXPORT TO PDF -->

The population of interest is all eligible voters. 

#### Part 2 ####

What is the sampling frame?

<!--
BEGIN QUESTION
name: q5b
manual: true
points: 1
-->
<!-- EXPORT TO PDF -->

The sampling frame is people who actually voted.  

### How might the sampling frame differ from the population?

After the fact, many experts have studied the 2016 election results. For example, according to the American Association for Public Opinion Research (AAPOR), predictions made before the election were flawed for three key reasons:

1. voters changed their preferences a few days before the election
2. those sampled were not representative of the voting population, e.g., some said that there was an overrepresentation of college graduates in some poll samples 
3. voters kept their support for Trump to themselves (hidden from the pollsters)

In the next two problems on this homework, we will do two things:

+ HW Question 6: We will carry out a study of the sampling error when there is no bias. In other words, we will try to compute the chance that we get the election result wrong even if we collect our sample in a manner that is completely correct. In this case, any failure of our prediction is due entirely to random chance.
+ HW Question 7: We will carry out a study of the sampling error when there is bias of the second type from the list above. In other words, we will try to compute the chance that we get the election result wrong if we have a small systematic bias. In this case, any failure of our prediction is due to a combination of random chance and our bias.


### Question 6

Why can't we assess the impact of the other two biases (voters changing preference and voters hiding their preference)? 

Note: You might find it easier to complete this question after you've completed the rest of the homework including the simulation study.


<!-- EXPORT TO PDF -->

You overestimate/underestimate the number of people actually planning to vote for each candidate.

### How large was the sampling error?

In some states the race was very close, and it may have been simply sampling error, i.e., random chance that the majority of the voters chosen for the sample voted for Clinton. 

One year after the 2016 election, Nate Silver wrote  in
*The Media Has A Probability Problem* that
the "media’s demand for certainty -- and its lack of statistical rigor -- is a bad match for our complex world."
FiveThirtyEight forecasted that Clinton had about a 70 percent chance of winning.  


A 2- or 3-point polling error in Trump’s favor (typical error historically)  would likely be enough to tip the Electoral College to him. 

We will first carry out a simulation study to assess the impact of the sampling error on the predictions.


## The Electoral College

The US president is chosen by the Electoral College, not by the
popular vote. Each state is alotted a certan number of 
electoral college votes, as a function of their population.
Whomever wins in the state gets all of the electoral college votes for that state.

There are 538 electoral college votes (hence the name of the Nate Silver's site, FiveThirtyEight).

Pollsters correctly predicted the election outcome in 46 of the 50 states. 
For these 46 states Trump received 231 and Clinton received 232 electoral college votes.

The remaining 4 states accounted for a total of 75 votes, and 
whichever candidate received the majority of the electoral college votes in these states would win the election. 

These states were Florida, Michigan, Pennsylvania, and Wisconsin.

|State |Electoral College Votes|
| --- | --- |
|Florida | 29 |
|Michigan | 16 |
|Pennsylvania | 20 |
|Wisconsin | 10|

For Donald Trump to win the election, he had to win either:
* Florida + one (or more) other states
* Michigan, Pennsylvania, and Wisconsin


The electoral margins were very narrow in these four states, as seen below:


|State | Trump |   Clinton | Total Voters |
| --- | --- |  --- |  --- |
|Florida | 49.02 | 47.82 | 9,419,886  | 
|Michigan | 47.50 | 47.27  |  4,799,284|
|Pennsylvania | 48.18 | 47.46 |  6,165,478|
|Wisconsin | 47.22 | 46.45  |  2,976,150|

Those narrow electoral margins can make it hard to predict the outcome given the sample sizes that the polls used. 

---
## Simulation Study of the Sampling Error

Now that we know how people actually voted, we can carry
out a simulation study that imitates the polling.

Our ultimate goal in this problem is to understand the chance that we will incorrectly call the election for Hillary Clinton even if our sample was collected with absolutely no bias.

### Question 7

#### Part 1 

For your convenience, the results of the vote in the four pivotal states is repeated below:

|State | Trump |   Clinton | Total Voters |
| --- | --- |  --- |  --- |
|Florida | 49.02 | 47.82 | 9,419,886  | 
|Michigan | 47.50 | 47.27  |  4,799,284|
|Pennsylvania | 48.18 | 47.46 |  6,165,478|
|Wisconsin | 47.22 | 46.45  |  2,976,150|


Using the table above, write a function `draw_state_sample(N, state)` that returns a sample with replacement of N voters from the given state. Your result should be returned as a list, where the first element is the number of Trump votes, the second element is the number of Clinton votes, and the third is the number of Other votes. For example, `draw_state_sample(1500, "florida")` could return `[727, 692, 81]`. You may assume that the state name is given in all lower case.

You might find `np.random.multinomial` useful.

<!--
BEGIN QUESTION
name: q7a
manual: true
points: 5
-->
<!-- EXPORT TO PDF -->

In [9]:
def draw_state_sample(N, state):
    
    florida = [0.4902, 0.4782, 0.0316]
    michigan = [0.4750, 0.4727, 0.0523]
    pennsylvania = [0.4818, 0.4746, 0.0436]
    wisconsin = [0.4722, 0.4645, 0.0633]
    if (state == 'florida'):
        statesample = np.random.multinomial(N, florida)
    elif (state == 'michigan'):
        statesample = np.random.multinomial(N, michigan)
    elif (state == 'pennsylvania'):
        statesample = np.random.multinomial(N, pennsylvania)
    else:
        statesample = np.random.multinomial(N, wisconsin)
    
    return statesample

#### Part 2

Now, create a function `trump_advantage` that takes in a sample of votes (like the one returned by `draw_state_sample`) and returns the difference in the proportion of votes between Trump and Clinton. For example `trump_advantage([100, 60, 40])` would return `0.2`, since Trump had 50% of the votes in this sample and Clinton had 30%.

<!--
BEGIN QUESTION
name: q7b
manual: true
points: 5
-->
<!-- EXPORT TO PDF -->

In [10]:
def trump_advantage(voter_sample):
    total_voters = (voter_sample[0] + voter_sample[1] + voter_sample[2])
    advantage = voter_sample[0]/total_voters - voter_sample[1]/total_voters
    return advantage

#### Part 3

Simulate Trump's advantage across 100,000 simple random samples of 1500 voters for the state of Pennsylvania and store the results of each simulation in a list called `simulations`. 

That is, `simulations[i]` should be Trump's percentage advantage for the `i`th simple random sample.

<!--
BEGIN QUESTION
name: q7c
manual: true
points: 5
-->
<!-- EXPORT TO PDF -->

In [11]:
simulations = []
for i in range(100000):
    simulations.append(trump_advantage(draw_state_sample(1500, "pennsylvania")))
simulations  

[0.03533333333333333,
 0.011333333333333362,
 -0.04799999999999999,
 -0.009333333333333305,
 -0.04999999999999999,
 -0.03866666666666663,
 -0.02266666666666667,
 -0.0013333333333332975,
 0.02266666666666667,
 0.00666666666666671,
 0.03666666666666668,
 0.02999999999999997,
 0.010666666666666658,
 0.01933333333333337,
 0.05333333333333329,
 0.013333333333333364,
 0.03200000000000003,
 -0.01733333333333331,
 0.03866666666666663,
 0.01533333333333331,
 -0.004666666666666652,
 0.04199999999999998,
 0.020000000000000018,
 0.03200000000000003,
 0.040666666666666684,
 0.04933333333333334,
 0.03933333333333333,
 0.031333333333333324,
 0.03866666666666663,
 -0.001333333333333353,
 0.00933333333333336,
 -0.006666666666666654,
 0.05466666666666664,
 -0.008000000000000007,
 -0.018000000000000016,
 0.03200000000000003,
 0.033333333333333326,
 -0.023999999999999966,
 0.040000000000000036,
 0.03866666666666663,
 0.03866666666666663,
 -0.008000000000000007,
 -0.013333333333333364,
 -0.0080000000000000

#### Part 4

Make a histogram of the sampling distribution of Trump's percentage advantage in Pennsylvania. In order to do this, we'll make use of Altair.  Altair takes a `pandas` data frame as input.  As the fist line in the cell below we create a data frame with your simulations as the only column. The second column, zero, is introduced for adding a vertical line to your plot at x = 0. Call `alt.Chart` on `simulation_df` to get started.

Make sure to give your plot a title and add labels where appropriate. You should use the [`mark_bar`](https://altair-viz.github.io/gallery/simple_bar_chart.html#gallery-simple-bar-chart) function in your code.

Make sure to [include a title](https://altair-viz.github.io/user_guide/customization.html#adjusting-axis-labels) as well as appropriate [axis labels](https://altair-viz.github.io/user_guide/customization.html#adjusting-axis-labels). 

<!--
BEGIN QUESTION
name: q7d
manual: true
points: 7
-->
<!-- EXPORT TO PDF -->

In [23]:
simulation_df = pd.DataFrame({'simulations' : simulations, 'zero' : 0})

bar_chart = alt.Chart(simulation_df).mark_bar().encode(
    alt.X("simulations:Q", bin=True),
    y='count()',
).properties(
    title="Pennsylvania Trump Advantage Percentages"
).configure_title(fontSize=18).configure_axis(
    labelFontSize=14, # change axes label font size
    titleFontSize=16  # change axes title font size
)


## The code below adds a vertical red line at 0 to the plot.
vline = alt.Chart(simulation_df).mark_rule(color='red', strokeWidth=3).encode(x='zero:Q')
bar_chart + vline


ValueError: Objects with "config" attribute cannot be used within LayerChart. Consider defining the config attribute in the LayerChart object instead.

#### Part 5

Now write a function `trump_wins(N)` that creates a sample of N voters for each of the four crucial states (Florida, Michigan, Pennsylvania, and Wisconsin) and returns 1 if Trump is predicted to win based on these samples and 0 if Trump is predicted to lose.

Recall that for Trump to win the election, he must either:
* Win the state of Florida and 1 or more other states
* Win Michigan, Pennsylvania, and Wisconsin

<!--
BEGIN QUESTION
name: q7e
manual: true
points: 5
gradescope: show
-->
<!-- EXPORT TO PDF -->

In [ ]:
def trump_wins(N):
    florida = trump_advantage(draw_state_sample(N, "florida"))
    mich = trump_advantage(draw_state_sample(N, "michigan"))
    penny = trump_advantage(draw_state_sample(N, "pennysylvania"))
    wis = trump_advantage(draw_state_sample(N, "wisconsin"))
    truthfalse = 0
    if (florida > 0 and (mich > 0 or penny > 0 or wis > 0)):
        truthfalse = 1
        return truthfalse
    elif(mich > 0 and penny > 0 and wis > 0):
        truthfalse = 1
        return truthfalse
    else:
        return truthfalse

#### Part 6

If we repeat 100,000 simulations of the election, i.e. we call `trump_wins(1500)` 100,000 times, what proportion of these simulations predict a Trump victory? Give your answer as `percent_trump`.

This number represents the percent chance that a given sample will correctly predict Trump's victory *even if the sample was collected with absoutely no bias*. 

**Note: Many laypeople, even well educated ones, assume that this number should be 1. After all, how could a non-biased sample be wrong? This is the type of incredibly important intuition we hope to develop in you throughout this class and your future data science coursework.**

<!--
BEGIN QUESTION
name: q7f
manual: true
points: 3
-->
<!-- EXPORT TO PDF -->

In [ ]:
counter = 0
for i in range(100000):
    counter += trump_wins(1500)
percent_trump = counter/100000
percent_trump

We have just studied the sampling error, and found out 
our predictions might look if there was no bias in our 
sampling process. 
Essentially, we assumed that the people surveyed didn't change their minds, 
didn't hide who they voted for, and were representative
of those who voted on election day.

---
## Simulation Study of Selection Bias

According to an article by Grotenhuis, Subramanian, Nieuwenhuis, Pelzer and Eisinga (https://blogs.lse.ac.uk/usappblog/2018/02/01/better-poll-sampling-would-have-cast-more-doubt-on-the-potential-for-hillary-clinton-to-win-the-2016-election/#Author):

"In a perfect world, polls sample from the population of voters, who would state their political preference perfectly clearly and then vote accordingly."

That's the simulation study that we just performed. 


It's difficult to control for every source of selection bias.
And, it's not possible to control for some of the other sources of bias.

Next we investigate the effect of small sampling bias on the polling results in these four battleground states.  

Throughout this problem, we'll examine the impacts of a 0.5 percent bias in favor of Clinton in each state. Such a bias has been suggested because highly educated voters tend to be more willing to participate in polls.

### Question 8

Throughout this problem, adjust the selection of voters so that there is a 0.5% bias in favor of Clinton in each of these states. 

For example, in Pennsylvania Clinton received 47.46 percent of the votes and Trump 48.18 percent. Increase the population of Clinton voters to 47.46 + 0.5  percent and correspondingly decrease the percent of Trump voters. 

#### Part 1

Simulate Trump's advantage across 100,000 simple random samples of 1500 voters for the state of Pennsylvania and store the results of each simulation in a list called `biased_simulations`.

That is, `biased_simulation[i]` should hold the result of the `i`th simulation.

That is, your answer to this problem should be just like your answer from Question 7 part 3, but now using samples that are biased as described above.

<!--
BEGIN QUESTION
name: q8a
manual: true
points: 5
-->
<!-- EXPORT TO PDF -->

In [ ]:
def draw_biased_state_sample(N, state):
    florida = [0.4897, 0.4787, 0.0316]
    michigan = [0.4745, 0.4732, 0.0523]
    pennsylvania = [0.4813, 0.4751, 0.0436]
    wisconsin = [0.4717, 0.4650, 0.0633]
    if (state == 'florida'):
        statesample = np.random.multinomial(N, florida)
    elif (state == 'michigan'):
        statesample = np.random.multinomial(N, michigan)
    elif (state == 'pennsylvania'):
        statesample = np.random.multinomial(N, pennsylvania)
    else:
        statesample = np.random.multinomial(N, wisconsin)
    
    return statesample
    
biased_simulations = []
for i in range(100000):
    biased_simulations.append(trump_advantage(draw_state_sample(1500, "pennsylvania"))) 
biased_simulations

#### Part 2

Make a histogram of the new sampling distribution of Trump's advantage now using these biased samples. That is, your histogram should be the same as in Q8.4, but now using the biased samples.

Make sure to give your plot a title and add labels where appropriate.

<!--
BEGIN QUESTION
name: q8b
manual: true
points: 5
-->
<!-- EXPORT TO PDF -->

In [ ]:

biased_simulation_df = pd.DataFrame({'simulations' : biased_simulations, 'zero' : 0})

bar_chart_biased = alt.Chart(biased_simulation_df).mark_bar().encode(
    x = alt.X("simulations", axis = alt.Axis(title = "Advantages")),
    y = alt.Y("zero", axis = alt.Axis(title = "Count"))
).properties(
    title="Pennsylvania Trump Advantage Percentages"
).configure_title(fontSize=18).configure_axis(
    labelFontSize=14, # change axes label font size
    titleFontSize=16  # change axes title font size
)

vline = alt.Chart(biased_simulation_df).mark_rule(color='red', strokeWidth=3).encode(x='zero:Q')
bar_chart_biased + vline



#### Part 3

Compare the histogram you created in Q8.2 to that in Q7.4. 


<!-- EXPORT TO PDF -->

I couldn't get my histograms to work properly.  The zero in the y-axis of both charts is a placeholder.  I plan to go into office hours next week to get this sorted out.  

#### Part 4

Now perform 100,000 simulations of all four states and return the proportion of these simulations that result in a Trump victory. This is the same fraction that you computed in Q7.6, but now using your biased samples.

Give your answer as `percent_trump_biased`.

This number represents the chance that a sample biased 0.5% in Hillary Clinton's favor will correctly predict Trump's victory. You should observe that the chance is signficantly lower than with an unbiased sample, i.e. your answer in Q7.6.


<!--
BEGIN QUESTION
name: q8d
manual: true
points: 5
-->
<!-- EXPORT TO PDF -->

In [ ]:
def trump_wins_biased(N): 
    florida = trump_advantage(draw_state_sample(N, "florida"))
    mich = trump_advantage(draw_state_sample(N, "michigan"))
    penny = trump_advantage(draw_state_sample(N, "pennysylvania"))
    wis = trump_advantage(draw_state_sample(N, "wisconsin"))
    truthfalse = 0
    if (florida > 0 and (mich > 0 or penny > 0 or wis > 0)):
        truthfalse = 1
        return truthfalse
    elif(mich > 0 and penny > 0 and wis > 0):
        truthfalse = 1
        return truthfalse
    else:
        return truthfalse

counter = 0
for i in range(100000):
    counter += trump_wins_biased(1500)
percent_trump_biased = counter/100000
percent_trump_biased    

## Further Study


### Question 9

Would increasing the sample size have helped?

#### Part 1

Try a sample size of 5,000 and run 100,000 simulations of a sample with replacement. What proportion of the 100,000 times is Trump predicted to win the election in the unbiased setting? In the biased setting?

Give your answers as `high_sample_size_unbiased_percent_trump` and `high_sample_size_biased_percent_trump`.

<!--
BEGIN QUESTION
name: q9a
manual: true
points: 5
-->
<!-- EXPORT TO PDF -->

In [ ]:
counter1 = 0
for i in range(100000):
    counter1 += trump_wins(5000) 
high_sample_size_unbiased_percent_trump = counter1/100000

counter2 = 0
for i in range(100000):
    counter2 += trump_wins_biased(5000)
high_sample_size_biased_percent_trump = counter2/100000

print(high_sample_size_unbiased_percent_trump, high_sample_size_biased_percent_trump)

#### Part 2

What do your observations from part 1 say about the impact of sample size
on the sampling error and on the bias?   

Extra question for those who are curious: Just for fun, you might find it interesting to see what happens with even larger sample sizes (> 5000 voters) for both the unbiased and biased cases. Can you get them up to 99% success with sufficient large samples? How many? Why or why not? If you do this, include your observations in your answer.

In [ ]:
counter3 = 0
for i in range(100000):
    counter3 += trump_wins(5000) 
trump1 = counter3/100000

counter4 = 0
for i in range(100000):
    counter4 += trump_wins_biased(5000)
trump2 = counter4/100000

print(trump1, trump2)

Write your answer in the cell below.
<!--
BEGIN QUESTION
name: q9b
manual: true
points: 2
-->
<!-- EXPORT TO PDF -->

As sample size increases, the sampling error and bias decrease.  Once you increase the sample size past 31000, you get success rates of 99%

### Question 10

According to FiveThirtyEight: "... Polls of the November 2016 presidential election were about as accurate as polls of presidential elections have been on average since 1972."

When the margin of victory may be relatively small as it was in 2016, why don't polling agencies simply gather significantly larger samples to bring this error close to zero?

<!--
BEGIN QUESTION
name: q10
manual: true
points: 2
-->
<!-- EXPORT TO PDF -->

It's difficult to gather larger samples.  When polling agencies send out polls and post on sites, it really depends on who sees them.  An uninterested internet user, for example, might not fill out the poll.  Anyone who doesn't have access to internet may choose to not send the poll in by mail.  

## Submission

**Important**: To make sure the test cases run correctly, click `Kernel>Restart & Run All` and make sure all of the test cases are still passing. Doing so will submit your code for you. 

You may submit this assignment as many times as you'd like before the deadline.

**You must restart and run all cells before submitting. Otherwise, you may pass test cases locally, but not on our servers. We will not entertain regrade requests of the form, “my code passed all of my local test cases, but failed the autograder”.**

# Running Built-in Tests
1. All tests are in `tests` directory
1. Each python file in `tests` is a test
1. `grader.check('testname')` runs test `'testname'`, e.g. `'q1'`
1. `grader.check_all()` runs all visible tests

In [ ]:
# Run built-in checks
grader.check_all()

In [ ]:
# Generate pdf in classic notebook (does not work in JupyterLab)
import nb2pdf
nb2pdf.convert('hw2.ipynb')

# To generate pdf using command-line, run in terminal,
# nb2pdf hw2.ipynb

# Submission Checklist
1. Check filename is 'hw2.ipynb'
1. Save file to confirm all changes are on disk
1. Run *Kernel > Restart & Run All* to execute all code from top to bottom
1. Check `grader.check_all()` output
1. Save file again to write any new output to disk
1. Check generated pdf that all responses are displayed correctly
1. Submit to Gradescope